### PR TITLE
[API][Behat] configure tax categories as API resource with tests

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         | api
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| License         | MIT

Screenshot 2020-02-26 at 14 25 03

todo: bulk delete (i prepared contexts for this functionality but for now it doesn't work by api)

### CODE REVIEW

File: src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+declare(strict_types=1);
+
+namespace Sylius\Behat\Context\Api\Admin;
+
+use Behat\Behat\Context\Context;
+use Sylius\Behat\Client\ApiClientInterface;
+use Sylius\Component\Taxation\Model\TaxCategoryInterface;
+use Webmozart\Assert\Assert;
+
+final class ManagingTaxCategoriesContext implements Context
```
Comment (Tomanhez) [inline src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php]: Wrong order in all steps, please sort it and group by: given, when, then

File: src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    {
+        $this->client->delete('tax_category', $taxCategory->getCode());
+    }
+
+    /**
+     * @Then /^(this tax category) should no longer exist in the registry$/
+     */
+    public function thisTaxCategoryShouldNoLongerExistInTheRegistry(TaxCategoryInterface $taxCategory): void
+    {
+        Assert::false($this->isItemOnIndex('code', $taxCategory->getCode()));
```
Comment (Tomanhez) [inline src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php]: Add custom exception messages in true/false assertions : https://github.com/Sylius/Sylius/pull/11141#discussion_r384135754

File: src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    {
+        Assert::same($this->client->getError(), 'code: The tax category with given code already exists.');
+    }
+
+    /**
+     * @Then there should still be only one tax category with :element :code
+     */
+    public function thereShouldStillBeOnlyOneTaxCategoryWith(string $element, string $code): void
+    {
+        //todo
```
Comment (Tomanhez) [inline src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php]: todo 🎉

File: src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    /**
+     * @Then I should be notified that :element is required
+     */
+    public function iShouldBeNotifiedThatIsRequired(string $element): void
+    {
+        if ($element === 'name') {
+            Assert::same($this->client->getError(), 'name: Please enter tax category name. name: sylius.tax_category.name.min_length');
+        } else {
+            Assert::same($this->client->getError(), sprintf('%s: Please enter tax category %s.', $element, $element));
+        }
```
Comment (Tomanhez) [inline src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php]: Maybe we should use two if blocks?

File: src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    }
+
+    /**
+     * @Then tax category with :element :name should not be added
+     */
+    public function taxCategoryWithElementValueShouldNotBeAdded(string $element, string $name): void
```
Comment (Tomanhez) [inline src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php]:
```suggestion
    public function taxCategoryWithElementNameShouldNotBeAdded(string $element, string $name): void
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    {
+        Assert::false($this->isItemOnIndex($element, $name));
+    }
+
+    /**
+     * @Then I should see a single tax category in the list
+     */
+    public function iShouldSeeTaxCategoriesInTheList(): void
```
Comment (Tomanhez) [inline src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php]:
```suggestion
    public function iShouldSeeASingleTaxCategoryInTheList(): void
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+    public function __construct(ApiClientInterface $client)
+    {
+        $this->client = $client;
+    }
+
+    /**
+     * @When I delete tax category :taxCategory
+     */
+    public function iDeletedTaxCategory(TaxCategoryInterface $taxCategory): void
```
Comment (Tomanhez) [inline src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php]:
```suggestion
    public function iDeleteTaxCategory(TaxCategoryInterface $taxCategory): void
```

File: src/Sylius/Behat/Context/Api/Admin/NotificationContext.php [hunk 8]
```diff
@@ -34,4 +34,28 @@ public function iShouldBeNotifiedThatSomethingHappened(): void
...
+        Assert::true($this->client->isUpdateSuccessful(), 'Resource could not be edited');
+    }
+
+    /**
+     * @Then I should be notified that they have been successfully deleted
+     */
+    public function iShouldBeNotifiedThatTheyHaveBeenSuccessfullyDeleted(): void
+    {
+        //todo, bulk delete will be cover in separate PR
+    }
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/NotificationContext.php]: Is it needed now?

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/TaxCategory.xml [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+<resources xmlns="https://api-platform.com/schema/metadata"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="https://api-platform.com/schema/metadata https://api-platform.com/schema/metadata/metadata-2.0.xsd">
+
+    <resource
+        class="Sylius\Component\Taxation\Model\TaxCategory"
+        shortName="TaxCategory"
+    >
+
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/TaxCategory.xml]: Unnecessary blank line

File: src/Sylius/Bundle/ApiBundle/Resources/config/serializer/TaxCategory.xml [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        </attribute>
+        <attribute name="updatedAt">
+            <group>tax_category:read</group>
+        </attribute>
+        <attribute name="id">
+            <group>tax_category:read</group>
+        </attribute>
+        <attribute name="code">
+            <group>tax_category:read</group>
+            <group>tax_category:write</group>
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/serializer/TaxCategory.xml]:
```suggestion
            <group>tax_category:create</group>
```
to be consistent with product options

File: src/Sylius/Bundle/ApiBundle/Resources/config/serializer/TaxCategory.xml [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+            <group>tax_category:update</group>
+        </attribute>
+        <attribute name="description">
+            <group>tax_category:read</group>
+            <group>tax_category:write</group>
+            <group>tax_category:update</group>
+        </attribute>
+    </class>
+
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/serializer/TaxCategory.xml]: Unnecessary blank line

File: src/Sylius/Behat/Client/ApiPlatformClient.php [hunk 12]
```diff
@@ -89,6 +89,11 @@ public function update(): void
         $this->request('PUT', $this->request['url'], ['CONTENT_TYPE' => 'application/ld+json'], $content);
     }
 
+    public function delete(string $resource, string $id): void
+    {
+        $this->request('DELETE', sprintf('/new-api/%s/%s', $resource, $id ), []);
```
Comment (GSadee) [inline src/Sylius/Behat/Client/ApiPlatformClient.php]:
```suggestion
        $this->request('DELETE', sprintf('/new-api/%s/%s', $resource, $id), []);
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+        $this->client->update();
+
+        Assert::false($this->client->hasValue('code', 'NEW_CODE'), 'The code field with value NEW_CODE exist');
+    }
+
+    /**
+     * @Then /^(this tax category) name should be "([^"]+)"$/
+     * @Then /^(this tax category) should still be named "([^"]+)"$/
+     */
+    public function thisTaxCategoryNameShouldBe(TaxCategoryInterface $taxCategory, $taxCategoryName): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingTaxCategoriesContext.php]:
```suggestion
    public function thisTaxCategoryNameShouldBe(TaxCategoryInterface $taxCategory, string $taxCategoryName): void
```

File: src/Sylius/Behat/Resources/config/services/contexts/api.xml [hunk 14]
```diff
@@ -36,5 +36,8 @@
         <service id="sylius.behat.context.api.admin.notification" class="Sylius\Behat\Context\Api\Admin\NotificationContext">
             <argument type="service" id="Sylius\Behat\Client\ApiClientInterface" />
         </service>
+        <service id="sylius.behat.context.api.admin.managing_tax_categories" class="Sylius\Behat\Context\Api\Admin\ManagingTaxCategoriesContext">
+            <argument type="service" id="Sylius\Behat\Client\ApiClientInterface" />
+        </service>
```
Comment (GSadee) [inline src/Sylius/Behat/Resources/config/services/contexts/api.xml]: Services in this file could have the alphabetical order

**Comment by GSadee** [comment]
Thanks, Adam! :tada:

**Comment by lchrusciel** [comment]
Part of #11250

**Pull request merged**